Introduces support for instantiating packages from scmsync

hennevogel opened the pull request:
Instead of returning nil for instantiating packages that come from a scmsync'ed Project, check if they are known to the backend and return an in-memory, read only Package object.

Fixes #13176

danidoni commented on src/api/app/controllers/webui/packages/files_controller.rb:
```diff
@@ -17,7 +17,7 @@ def show
         @addeditlink = false
 
-        if User.possibly_nobody.can_modify?(@package) && @rev.blank? && @package.scmsync.blank?
+        if User.possibly_nobody.can_modify?(@package) && @rev.blank? && @package.scmsynced?
```
```suggestion
        if User.possibly_nobody.can_modify?(@package) && @rev.blank? && !@package.scmsynced?
```

danidoni commented on src/api/app/models/bs_request_action_submit.rb:
```diff
@@ -95,7 +95,7 @@ def execute_accept(opts)
     # cleanup source project
     if relink_source && sourceupdate != 'noupdate'
-      if Package.find_by_project_and_name(source_project, source_package).scmsync.blank?
+      if Package.find_by_project_and_name(source_project, source_package).scmsynced?
```
```suggestion
      unless Package.find_by_project_and_name(source_project, source_package).scmsynced?
```

hennevogel commented:
Do not merge this.

It needs some more sophisticated authorization setup. We probably want to switch the default of the option and then explicitly enable this per Package.get_by_project_and_name call and figure out problems first....

hennevogel commented:
I switched this off by default so we can start experimenting with it for instance with stuff like #15652

hennevogel committed:
Introduces support for instantiating packages from scmsync

Instead of returning nil for instantiating packages that come from a scmsync'ed
Project, check if they are known to the backend and return an in-memory, read only
Package object with what we know.

Co-authored-by: Adrian Schröter <adrian@suse.de>

adrianschroeter commented on src/api/app/models/package.rb:
```diff
@@ -184,8 +184,20 @@ def self.internal_get_project(project)
+  #
+  # You can follow this type of project link and try to find the Package from the "SCM" Project
+  # by setting in the opts hash:
+  #   follow_project_scmsync_links: true
```
please name it more neutral like "use_backend_only_packages: true" as this is also the case for any package from remote instances. So we should not be scmsync specific here.

hennevogel replied:
Your was, this is not. This is limited to scmsync.

krauselukas commented on src/api/app/models/package.rb:
```diff
@@ -196,7 +208,18 @@ def self.get_by_project_and_name(project_name_or_object, package_name, opts = {}
     return unless project # remote prjs
 
-    return nil if project.scmsync.present?
+    if project.scmsync.present? && opts[:follow_project_scmsync_links]
```
@hennevogel the intention is to keep the current behavior unless `opts[:follow_project_scmsync_links]` is set to true by hand right? at this place we used to check for `project.scmsync.present?` and returned `nil` otherwise. I think to make sure we still return `nil` in other cases it need a condition `return nil if project.scmsync.present? && !opts[:follow_project_scmsync_links]` after this block.
